src/postgres_deparse.c

```diff
@@ -130,173 +94,583 @@ isOp(const char *val)
+					deparseAppendSubpart(state);
+			}
+			else
+			{
+				appendStringInfoChar(str, '\n');
+			}
+		}
+
+		if (state->current) /* refresh to handle new subpart(s) */
+			str = deparseGetCurrentStringInfo(state);
+		appendStringInfoString(str, comment->str);
+
+		for (int j = 0; j < comment->newlines_after_comment; j++)
+		{
+			if (state->current && state->opts.pretty_print)
+				deparseAppendSubpart(state);
+			else
+				appendStringInfoChar(str, '\n');
+		}
+
+		state->emitted_comments = bms_add_member(state->emitted_comments, i);
+	}
+}
+
+static void
+deparseEmit(DeparseState *state, StringInfo str)
+{
+	ListCell *lc;
+	foreach (lc, state->result)
+	{
+		DeparseStateSubPart *subpart = (DeparseStateSubPart *) lfirst(lc);
+		bool first_subpart = list_cell_number(state->result, lc) == 0;
+		bool last_subpart = list_cell_number(state->result, lc) == list_length(state->result) - 1;
+
+		if (state->opts.pretty_print && !first_subpart && (!last_subpart || subpart->str->len > 0))
+		{
+			appendStringInfoChar(str, '\n');
+		}
+		else if (!first_subpart && (state->opts.pretty_print || str->data[str->len - 1] != '('))
+		{
+			appendStringInfoChar(str, ' ');
+		}
+
+		if (state->opts.pretty_print)
+		{
+			for (int i = 0; i < subpart->indent; i++)
+				appendStringInfoChar(str, ' ');
+		}
+		appendStringInfoString(str, subpart->str->data);
+		removeTrailingSpaceFromStr(str);
+
+		pfree(subpart->str->data);
+		pfree(subpart->str);
+		pfree(subpart);
+	}
+	list_free(state->result);
+	state->result = NIL;
+	if (state->opts.pretty_print && state->opts.trailing_newline)
+		appendStringInfoChar(str, '\n');
+}
+
+static void
+deparseStatePushStatement(DeparseState *state)
+{
+	DeparseStateStatement *stmt = palloc0(sizeof(DeparseStateStatement));
+	stmt->parent = state->current;
+	if (stmt->parent)
+	{
+		DeparseStatePart *part = deparseGetCurrentPart(state);
+		stmt->base_indent = stmt->parent->base_indent + state->opts.indent_size;
+		if (part->indent_subparts) /* Indent again if subparts next to us are also indented */
+			stmt->base_indent += state->opts.indent_size;
+	}
+	state->current = stmt;
+}
+
+static void
+deparseStatePopStatement(DeparseState *state)
+{
+	ListCell *lc;
+	ListCell *lc2;
+	DeparseStateStatement *stmt = state->current;
+	Assert(stmt != NULL);
+	List **result = NULL;
+
+	if (stmt->parent)
+	{
+		state->current = stmt->parent;
+		result = &deparseGetCurrentPart(state)->subparts;
+	}
+	else
+	{
+		// If its the top-level statement, save as results instead
+		result = &state->result;
+	}
+
+	// Step through all lines, as if we did the final emit, but instead emit line-by-line into the parent
+	foreach (lc, stmt->parts)
+	{
+		DeparseStatePart *part = (DeparseStatePart *) lfirst(lc);
+		bool first_part = list_cell_number(stmt->parts, lc) == 0;
+
+		/* Only merge subparts if they all have the same indentation level */
+		/* TODO: We can be a bit more permissive here - basically we want to avoid merging across statement levels */
```

Comment:
To support this better (e.g. for the `complex_gitlab.sql` test case), we could also alternatively consider keeping a boolean override on the subpart that indicates that the part shall not be merged and must stay on its own line (vs piggybacking on the indent here).